server/src/main/java/org/apache/iotdb/db/engine/flush/MemTableFlushTask.java

```diff
@@ -218,159 +265,59 @@ public void syncFlushMemTable() throws ExecutionException, InterruptedException
-              ((IChunkWriter) ioMessage).writeToFileWriter(this.writer);
-            }
-          } catch (IOException e) {
-            LOGGER.error(
-                "Database {} memtable {}, io task meets error.", storageGroup, memTable, e);
-            throw new FlushRunTimeException(e);
-          }
-          long subTaskTime = System.currentTimeMillis() - starTime;
-          ioTime += subTaskTime;
-          WRITING_METRICS.recordFlushSubTaskCost(WritingMetrics.IO_TASK, subTaskTime);
-        }
-        LOGGER.debug(
-            "flushing a memtable to file {} in database {}, io cost {}ms",
-            writer.getFile().getName(),
-            storageGroup,
-            ioTime);
-        WRITING_METRICS.recordFlushTsFileSize(storageGroup, writer.getFile().length());
-        WRITING_METRICS.recordFlushCost(WritingMetrics.FLUSH_STAGE_IO, ioTime);
-      };
-
-  static class TaskEnd {
+  protected void metricFlush() {
+    if (!storageGroup.startsWith(IoTDBMetricsUtils.DATABASE)) {
+      int lastIndex = storageGroup.lastIndexOf("-");
+      if (lastIndex == -1) {
+        lastIndex = storageGroup.length();
+      }
+      MetricService.getInstance()
+          .gaugeWithInternalReportAsync(
+              memTable.getTotalPointsNum(),
+              Metric.POINTS.toString(),
+              MetricLevel.CORE,
+              Tag.DATABASE.toString(),
+              storageGroup.substring(0, lastIndex),
+              Tag.TYPE.toString(),
+              "flush",
+              Tag.REGION.toString(),
+              dataRegionId);
+    }
+  }
 
-    TaskEnd() {}
+  private DynamicThread newSortThread() {
+    return new TaskRunner(
+        sortTasks, this::cleanSortThread, sortTaskQueue, encodingTaskQueue, taskName + "-sort");
   }
 
-  static class EndChunkGroupIoTask {
+  private DynamicThread newEncodingThread() {
+    return new TaskRunner(
+        encodingTasks,
+        this::cleanEncodingThread,
+        encodingTaskQueue,
+        ioTaskQueue,
+        taskName + "-encode");
+  }
 
-    EndChunkGroupIoTask() {}
+  private DynamicThread newIOThread() {
+    return new TaskRunner(null, this::cleanIOThread, ioTaskQueue, ioTaskQueue, taskName + "-io");
   }
 
-  static class StartFlushGroupIOTask {
+  private void cleanSortThread() {
+    metricFlush();
+    WRITING_METRICS.recordFlushCost(
+        WritingMetrics.FLUSH_STAGE_SORT, allContext.getSortTime().get());
+  }
 
-    private final String deviceId;
+  private void cleanEncodingThread() {
+    metricFlush();
+    WRITING_METRICS.recordFlushCost(
+        WritingMetrics.FLUSH_STAGE_ENCODING, allContext.getEncodingTime().get());
+  }
 
-    StartFlushGroupIOTask(String deviceId) {
-      this.deviceId = deviceId;
-    }
+  private void cleanIOThread() {
+    metricFlush();
+    WRITING_METRICS.recordFlushCost(WritingMetrics.FLUSH_STAGE_IO, allContext.getIoTime().get());
+    WRITING_METRICS.recordFlushTsFileSize(storageGroup, allContext.getWriter().getFile().length());
   }
```

Comment:
why call` metricFlush() `after each task pipeline?

Reply:
The original code did this at the end of each stage. I am not sure of its functionality, so I keep it.

Reply:
I see the previous code implementation only calls metricFlush at the end of the entire flush task. Maybe you need to confirm.

Reply:
Now I only retain the one in the IOTask.